### ISSUE TITLE
Refactor extension for translation

### ISSUE THREAD

**Issue by thomaschampagne** [thomaschampagne/StravaPlus#10]

**Comment by praveentiru** [comment]
Documenting approach for supporting translations:

Current Status:
* Chrome supports all locales that are supported by the [Chrome web store](https://developer.chrome.com/webstore/i18n?csw=1#localeTable)
* Content scripts cannot access chrome.i18n.* methods needed to support translation
* Moreover, we cannot request permission for these apis as well - List [HERE](https://developer.chrome.com/extensions/declare_permissions)
* Need to declare the translation service as background and build communication with display page

- [ ] Implement small PoC to prove the concept - Use the stravistix menu as testing ground
- [ ] Iterate if the mechanism is ok for wider implementation
- [ ] Implement a production ready method with mechanism

Please document if you see any other alternatives.

**Comment by praveentiru** [comment]
@thomaschampagne. After a lot of pain and suffering, I have finally basic code in place to run translation. In the screenshot below, you are seeing ####### in the menu because they are translation strings for en-GB. Hopefully, things would be easier now.

@thomaschampagne I will push changes to my fork. You can look at them there if you are interested. Would love feedback.

![change_menu_text](https://cloud.githubusercontent.com/assets/12044936/15621193/de567432-247c-11e6-9076-484bec270152.PNG)

**Comment by praveentiru** [comment]
Step 1: Enable translation for messages added to view through Javascript

- [ ] Implement PoC
- [ ] Implement String substituion - [Description](https://github.com/jquery/globalize/blob/master/doc/api/message/message-formatter.md#simple-variable-replacement)
- [ ] Implement Message Inheritance - [Description](https://github.com/jquery/globalize/blob/master/doc/api/message/load-messages.md#messages-inheritance)
- [ ] Develop Helper method to run translations
- [ ] Enable translation across all js pages

Step 2: Implement Number formatting (We live different worlds: 1,00,000 or 100,000 or 100.000)

- [ ] Develop Helper method for number formatting
- [ ] Implement changes to js to ensure proper formatting of numbers

Step 3: What to do about angular pages?

[This](http://blog.trifork.com/2014/04/10/internationalization-with-angularjs/) clearly explains why basic angular i18n is not good enough. Need to see if using something like [angular-translate](https://github.com/angular-translate/angular-translate) is better approach rather than re-using jQuery Globalize.

**Comment by praveentiru** [comment]
Translation enabled for JS views under common section of Stravistix.js. Two minor changes pending due to absence of app resources in modifiers
* SegmentRankPercentageModifier
* DisplayFlyByFeedModifier

**Comment by praveentiru** [comment]
@thomaschampagne documenting challenge here so that all have access

I ran into trouble when I saw worker thread for activity analysis was failing to execute. I have finally isolated the source of issue. **It is very strange**.

The following code triggers the problem - Code [here](https://github.com/praveentiru/stravistix/blob/9a478e89268b6809a6bde2a38a1353aa6f878848/hook/extension/js/StravistiX.js#L63-L82)
```javascript
$.getJSON(this.appResources_.cldrBase, function (data) {
   // #10 - Loading CLDR data needed for globalize to work
   Globalize.load(data);
   // #10 - Loading translation libraries from locales folder
   var loadCount = 0;
   for (var i = 0; i < locArray.length; i++) {
        $.getJSON(locArray[i], function (data) {
        Globalize.loadMessages(data);
        loadCount++;
        if (loadCount === locArray.length) {
               var currentLocale = window.navigator.language || window.navigator.userLanguage;
               // We should create a global instance for translation only after loading all messages
               appRes.globalizeInstance = Globalize(currentLocale);
                // #10 - Moving all the UI generation into function to ensure globalize instance
                // is properly built as we need it to translate
                strav.handleApp_();
            }
        });
    }
});
```

strav.handleApp_ builds the UI of the extension. I have structured the code so that, it is triggered only after all translation libraries are loaded properly which would enable me to translate the UI as it is getting built. This method of building UI is triggering failure of worker in [ActivityProcessor.js@Line 110](https://github.com/thomaschampagne/stravistix/blob/2433488acd156c4d3ed226534cc850e4e4d7f36f/hook/extension/js/processors/ActivityProcessor.js#L110). Web worker fails due to inability to clone the object.

This problem does not happen handleApp_ call is moved out of callback method. Advice on how to resolve this would be great.

**Comment by praveentiru** [comment]
Found the culprit. It is `appRes.globalizeInstance`.

One of the input variables for Web Worker is appResources_. I have added `globalizeInstance` to appResources_ to enable access to translator across the app. Web Worker is unable to clone this object and it fails. The reason it works when we move out of scope of callback is because web worker starts before instance of globalize could be instantiated.

Now, I will look for solution :)

**Comment by praveentiru** [comment]
Fix to solve this issue.

Change 1@[AcitivityProcessor.js#L113](https://github.com/thomaschampagne/stravistix/blob/2433488acd156c4d3ed226534cc850e4e4d7f36f/hook/extension/js/processors/ActivityProcessor.js#L113)
Change `appResources: this.appResources,` to `extnId: this.appResources.extensionId,`

Change 2@[ComputeAnalysisWorker.js#L31](https://github.com/thomaschampagne/stravistix/blob/2433488acd156c4d3ed226534cc850e4e4d7f36f/hook/extension/js/processors/workers/ComputeAnalysisWorker.js#L31)
Change `self.importRequiredLibraries(self.required, mainThreadEvent.data.appResources.extensionId);` to ' self.importRequiredLibraries(self.required, mainThreadEvent.data.extnId);'

Web worker does not need everything from appresources only extension id to load javascripts. :)

**Comment by praveentiru** [comment]
Just uploaded a large commit - [Here](https://github.com/praveentiru/stravistix/commit/9d808a1e7179beb77fa62834a35f90c4eb75b491).

This completes creation of default message library for summary of extended statistics (Not segment only activity) and detailed running stats. Sample screenshot with locale set to en-GB. (Using ###### as translation strings for testing).

![remoteview extended summary](https://cloud.githubusercontent.com/assets/12044936/15633010/2e9500b4-25c0-11e6-80e4-0a16745e4591.PNG)
![detailed running stats](https://cloud.githubusercontent.com/assets/12044936/15633012/35d7288e-25c0-11e6-8631-335fdef69744.PNG)

**Comment by thomaschampagne** [comment]
Very nice ! Great job !

I merged your code via pull request into a **feature/10** on my repo. I try to support french at the moment. I did't fell into fallback language (eg. en-US) at this time i getting this:

```
globalize.js:105 Uncaught Error: E_MISSING_MESSAGE_BUNDLE: Missing message bundle for locale `fr`.
createError	@	globalize.js:105
validate	@	globalize.js:182
validateMessageBundle	@	message.js:1873
Globalize.messageFormatter.Globalize.messageFormatter	@	message.js:2031
Globalize.formatMessage.Globalize.formatMessage	@	message.js:2068
Helper.translateDOMNode	@	Helper.js:206
modify	@	MenuModifier.js:55
handleMenu_	@	StravistiX.js:296
handleApp_	@	StravistiX.js:90
(anonymous function)	@	StravistiX.js:77
c	@	strava-head-15cde5e….js:24
fireWith	@	strava-head-15cde5e….js:24
n	@	strava-head-15cde5e….js:25
(anonymous function)	@	strava-head-15cde5e….js:25
```

Currently discovering and learning from your job ;)

**Comment by praveentiru** [comment]
Message inheritance is properly implemented. So, you need not copy all the message keys to fr.json file. Just keep the one you have translated and remove others. Translation will fall back to root.json

I am pushing a fix where extension fails when the file is not supported. It should not fail like it id.

**Comment by thomaschampagne** [comment]
@praveentiru Indeed, i didn't need to override all keys in french if no translations root file should do the work :) Just copy past file :) Hihi ^^^

**FYI** Here is locales list from language dropdown list of strava:
```
en-GB
de-DE
en-US
es-ES
es-41
fr-FR
it-IT
nl-NL
pt-PT
pt-BR
ru-RU
ko-KR
zh-CN
ja-JP
zh-TW
```
I mean that list: 
![image](https://cloud.githubusercontent.com/assets/151973/15634323/50d264ea-25c1-11e6-869c-0bb35b27a1ca.png)

And chrome locale list is: https://developer.chrome.com/webstore/i18n#localeTable

Locales value you got if you type this in a chrome console:
```
window.navigator.language
```

Thanks !!!!!!!

**Comment by praveentiru** [comment]
Actually, you do not have to copy paste the root.json. In your fr.json remove everything except for code below:
```
{
    "fr": {
        "menu": {
            "common_settings": "Paramètres principaux",
            "health_settings": "Paramètres de santé"
        }
    }
}
```
You will see that rest of UI will reflect the strings in root.json. This way it is much more efficient and translator can work on adding one section at a time. :)

Now, you need to search for volunteers to support all those languages after release. :)

**Comment by thomaschampagne** [comment]
@praveentiru 

> You will see that rest of UI will reflect the strings in root.json. This way it is much more efficient and translator can work on adding one section at a time. :)

Sure !

> Now, you need to search for volunteers to support all those languages after release. :)

Should not be a problem ;)

**Comment by thomaschampagne** [comment]
@praveentiru 

* I'm getting this error with last PR. I'm searching for understand and fixing now
```
globalize.js:105 Uncaught Error: E_MISSING_MESSAGE_BUNDLE: Missing message bundle for locale `en-US`.
````

* We should cover only locales supported by strava:

 >en-GB, de-DE, en-US, es-ES, es-41, fr-FR, it-IT, nl-NL, pt-PT, pt-BR, ru-RU, ko-KR, zh-CN, ja-JP, zh-TW

Don't need to translate a language like Greek not support by strava in their web interface

Then have:

```javascript
supportedLocales: [ 'root', 'en-GB', 'de-DE', 'en-US', 'es-ES', 'es-41', 'fr-FR', 'it-IT', 'nl-NL', 'pt-PT', 'pt-BR', 'ru-RU', 'ko-KR', 'zh-CN', 'ja-JP', 'zh-TW' ];
```
Do you confirm?

**Note:** strava store the locale used by user into a cookie with key named **ui_language**
![image](https://cloud.githubusercontent.com/assets/151973/15635298/14a744da-25db-11e6-9b45-451334933f64.png)


* About date formatting you listed i introduced moment.js in **feature/218(-poc)|213(-poc)** 
Lot of new things mixed once :) Maybe delay this in that branch.. i can cover it easily localized date formats later with **moment.js** (already managed in a pro project thought 10 languages ;)).

**Comment by praveentiru** [comment]
@thomaschampagne I just tested with my feature/pty-#10 branch code which was my last Pull request and it is working fine. Not able to re-produce. That error should not occur because en-US.json is present. Can you try getting callstack when you are getting the error? Also, what was the chrome locale when it happened.

I agree that we should support only the locales supported by Strava and pull the locale used by strava. If strava supports dynamically changing language then, I need to see how my mechanism works.

I will focus on textual translation for now. Will look into date translation if needed at a later date.

**Comment by praveentiru** [comment]
Some screenshots after the latest commit. As you can see, labels of chartjs need to enabled for translation. Otherwise all text elements have been addressed.

![extended stats](https://cloud.githubusercontent.com/assets/12044936/15665448/dd8b463c-2728-11e6-87f0-aaf46c49b5a6.PNG)
![athlete summary activity summary](https://cloud.githubusercontent.com/assets/12044936/15665454/e1befb22-2728-11e6-8966-199ae09a16f9.PNG)
![athlete stats chart](https://cloud.githubusercontent.com/assets/12044936/15665457/e8497fee-2728-11e6-83de-bd5b79afd591.PNG)

**Comment by praveentiru** [comment]
@thomaschampagne PoC for enabling translation in angular app used for settings up and running. YAAY!!! Look at the beautiful hashes. :) :)

Re-using globalize itself. Created a new service by name TranslationService to support translation in angular app.

![image](https://cloud.githubusercontent.com/assets/12044936/15735682/2d6c5598-28b8-11e6-976e-3be54cb5dcda.png)

**Comment by praveentiru** [comment]
@thomaschampagne We need to discuss if the current progress is enough to add it to release. It would give a good baseline to look for volunteers to start translating. In the next release, we can add the missing elements.

Your thoughts.
